pjsip/include/pjsua2/endpoint.hpp

```diff
@@ -437,6 +437,150 @@ struct OnMediaEventParam
     MediaEvent      ev;
 };
 
+/**
+ * This structure describes authentication challenge used in Proxy-Authenticate
+ * or WWW-Authenticate for digest authentication scheme.
+ */
+struct DigestChallenge
+{
+    /**
+     * Realm for the challenge.
+     * */
+    std::string realm;
+
+    /*
+     * Other parameters.
+     * */
+    std::map<std::string, std::string> otherParam;
+
+    /**
+     * Domain.
+     */
+    std::string domain;
+
+    /**
+     * Nonce challenge.
+     */
+    std::string nonce;
+
+    /**
+     * Opaque value.
+     */
+    std::string opaque;
+
+    /**
+     * Stale parameter.
+     */
+    int  stale;
+
+    /**
+     * Algorithm parameter.
+     */
+    std::string algorithm;
+
+    /**
+     * Quality of protection.
+     */
+    std::string qop;
+
+    /**
+     * Convert from pjsip
+     */
+    void fromPj(const pjsip_digest_challenge &prm);
+
+    /**
+     * Convert to pjsip
+     */
+    pjsip_digest_challenge toPj() const;
+};
+
+/**
+ * This structure describe credential used in Authorization and
+ * Proxy-Authorization header for digest authentication scheme.
+ */
+struct DigestCredential
+{
+    /**
+     *Realm of the credential
+     */
+    std::string realm;
+
+    /**
+     *Other parameters.
+     */
+    std::map<std::string, std::string> otherParam;
```

Comment:
re: `std::map<std::string, std::string>`, for SWIG interfacing purpose, better to use predefined type (similar to `StringVector`) in `types.hpp`, e.g:
```
/** Map string to string */
typedef std::map<std::string, std::string> StringToStringMap;
```

and then in `pjsip-apps/src/swig/pjsua.i`, add (see lines started with `+`):
```
//
// STL stuff.
//
%include "std_string.i"
%include "std_vector.i"
+ %include "std_map.i"

%template(StringVector)			std::vector<std::string>;
%template(IntVector) 			std::vector<int>;
+ %template(StringToStringMap)		map<string, string>;
```